builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	switch role.Generate {
+	case true:
+		key_object, err := otplib.NewKeyFromURL(role.URL)
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a Key object."), nil
+		}
+
+		barcode, err := key_object.Image(role.QRSize, role.QRSize)
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a QR code image."), nil
+		}
+
+		var buff bytes.Buffer
+		png.Encode(&buff, barcode)
+		b64Barcode := base64.StdEncoding.EncodeToString(buff.Bytes())
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"url":     key_object.String(),
+				"barcode": b64Barcode,
+				"skew":    role.Skew,
+				"qr_size": role.QRSize,
+			},
+		}, nil
+	default:
+		// Translate algorithm back to string
+		algorithm := role.Algorithm.String()
+
+		// Return values of key
+		return &logical.Response{
+			Data: map[string]interface{}{
+				"issuer":       role.Issuer,
+				"account_name": role.AccountName,
+				"period":       role.Period,
+				"algorithm":    algorithm,
+				"digits":       role.Digits,
+			},
+		}, nil
+	}
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("key/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	generate := data.Get("generate").(bool)
+	key_string := data.Get("key").(string)
+	issuer := data.Get("issuer").(string)
+	account_name := data.Get("account_name").(string)
+	period := data.Get("period").(int)
+	algorithm := data.Get("algorithm").(string)
+	digits := data.Get("digits").(int)
+	skew := data.Get("skew").(int)
+	qr_size := data.Get("qr_size").(int)
+	key_size := data.Get("key_size").(int)
+
+	// Translate digits and algorithm to a format the totp library understands
+	var role_digits otplib.Digits
+	switch digits {
+	case 6:
+		role_digits = otplib.DigitsSix
+	case 8:
+		role_digits = otplib.DigitsEight
+	default:
+		return logical.ErrorResponse("The digit value can only be 6 or 8."), nil
+	}
+
+	var role_algorithm otplib.Algorithm
+	switch algorithm {
+	case "SHA1":
+		role_algorithm = otplib.AlgorithmSHA1
+	case "SHA256":
+		role_algorithm = otplib.AlgorithmSHA256
+	case "SHA512":
+		role_algorithm = otplib.AlgorithmSHA512
+	default:
+		return logical.ErrorResponse("The algorithm value is not valid."), nil
+	}
+
+	// Enforce input value requirements
+	if period <= 0 {
+		return logical.ErrorResponse("The period value must be greater than zero."), nil
+	}
+
+	if skew < 0 {
```

Comment:
You may want to ensure that `skew` isn't greater than 1. There is unlikely to ever be a reason you need a skew greater than 1 period on either end.